src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5874,6 +5883,66 @@ private ScanTypeFlags ScanTupleType(out SyntaxToken lastTokenOfType)
             return ScanTypeFlags.NotType;
         }
 
+#nullable enable
+        private ScanTypeFlags ScanFunctionPointerType(out SyntaxToken lastTokenOfType)
+        {
+            Debug.Assert(IsFunctionPointerStart());
+            _ = EatContextualToken(SyntaxKind.DelegateKeyword);
+            lastTokenOfType = EatToken(SyntaxKind.AsteriskToken);
+
+            if (IsPossibleCallingConvention())
+            {
+                _ = EatToken();
+            }
+
+            if (!IsPossibleFunctionPointerParameterListStart(CurrentToken))
+            {
+                return ScanTypeFlags.MustBeType;
```

Comment:
>ScanTypeFlags.MustBeType [](start = 23, length = 24)

It feels like the situation doesn't quite fit the meaning of the value:
```
/// <summary>
/// Definitely a type name: either a predefined type (int, string, etc.) or an array
/// type (ending with a [] brackets), or a pointer type (ending with *s).
/// </summary>
MustBeType,
```

Also, should we somehow indicate that the type is syntactically incomplete?

Reply:
I talked with Neal about this during the initial passes. We use ScanTypeFlags for later syntax disambiguation, and `delegate*` is 100% definitely a type. I'll go through existing usages of the flag to make sure that this doesn't add new ambiguities, and update the documentation for the invalid cases.

---
In reply to: [349718535](https://github.com/dotnet/roslyn/pull/39879#discussion_r349718535) [](ancestors = 349718535)